### ISSUE TITLE
unexpected results in inference with increased maximum sequence length

### ISSUE THREAD

**Issue by Louis-y-nlp** [mosaicml/llm-foundry#331]
## ❓ Question
```python
import transformers

name = 'mosaicml/mpt-7b-instruct'

config = transformers.AutoConfig.from_pretrained(name, trust_remote_code=True)
config.max_seq_len = 4096 # (input + output) tokens can now be up to 4096

model = transformers.AutoModelForCausalLM.from_pretrained(
  name,
  config=config,
  trust_remote_code=True
)
```
I'm using the recommended approach by HF to increase the length using Alibi, but I'm getting really poor results. It can't even correctly tell me what comes after 2000; I get empty string outputs or meaningless results。

## Additional context

here are some examples ：
input:    "please tell me what is the next number ?" + ' '.join([str(i) for i in range(1000)])
output: 'The next number after 0,1,2,3...999 is 1000'

input: "please tell me what is the next number ?" + ' '.join([str(i) for i in range(2000)])
output:  ''

input: "please tell me what is the next number?" + ' '.join([str(i) for i in range(2000)])
output: "The first thing you can do to help your own personalised with guidance and support services, so we will be able to make sure it's safe for our favourite pet dog in this section of course 0.001]]. The most important part of my life was born as well but also have some people who are not only because they're very helpful when I am going to use their ability to get back into the world class[.]\nThe following were written by using the same.[]\n\n  the best way to start getting ready to take advantage to find out how much time spent on the process which could be found at least [the value]'. This would like to know about the latest reports from being too long ago given to check rules over the decision to stay away from the possibility to look up to spend the last decisions to attend without having been made available to rule or any other words)\n\n## What’s happening today?”\n\n#### To improve Our mission statement “This”\n\n## A few months later.”\n\n## You may need to read the text below\n\n## And finally decided to write something else\n\n## That has been done\n\n## There are no longer\n\n## An example of the city\n\n## It should be obvious\n\nTo put them\n\nA manhattan\n\nYou might Notional\n\nIf you don’t want to work\n\nAncient\n\n“It’Selling\n\nIn general, there are six different ways to attractiveness\n\nI think writing\n\nIsolated IncognitoESCAPEPEACEUVICECTIONDESKTOPICIPEDESPECTIVE DESIGNATESpecialized DesigneespecializeDesignDesignedEspecially DesignedespecializeDESIGNEDESPECTDESTINE DESCRIPTIONDESIREDESCRIPTEDESPECTDESINE DESIGENTSpecializingDESIGNESSIPDESIREDESPECTINGDESIGNESPECTDESINE DESIDESpecializespecialitiespecialismspecializedDESIGNSDESIDESTDNDESILDESINEIDSpectingDESIGNESPECTIMERSpecialismidesimpleDESIGNESPECTIMEISLESpecializedDESDESMANDESpecializedDESETDESINELLEDESPECITYDESERVICESpecializedDESERTDESINEUXESpecificatespecialisespecialsitespeciousdeservespecitingDESECTDESULTESpecialitiespecialistencespecitableDESDDESPECTDESINEATTRACTED"

here is my python script

```python
from typing import Any, Dict, Tuple
import warnings

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig
from transformers import StoppingCriteria, StoppingCriteriaList


INSTRUCTION_KEY = "### Instruction:"
RESPONSE_KEY = "### Response:"
END_KEY = "### End"
INTRO_BLURB = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
PROMPT_FOR_GENERATION_FORMAT = """{intro}
{instruction_key}
{instruction}
{response_key}
""".format(
    intro=INTRO_BLURB,
    instruction_key=INSTRUCTION_KEY,
    instruction="{instruction}",
    response_key=RESPONSE_KEY,
)


class InstructionTextGenerationPipeline:
    def __init__(
        self,
        model_name,
        torch_dtype=torch.bfloat16,
        trust_remote_code=True,
        use_auth_token=None,
    ) -> None:
        config = AutoConfig.from_pretrained(model_name, trust_remote_code=True)
        config.max_seq_len = 4096
        self.model = AutoModelForCausalLM.from_pretrained(
            model_name,
            torch_dtype=torch_dtype,
            use_auth_token=use_auth_token,
            config = config,
            trust_remote_code=True,
        )

        tokenizer = AutoTokenizer.from_pretrained(
            model_name,
            trust_remote_code=trust_remote_code,
            use_auth_token=use_auth_token,
        )
        if tokenizer.pad_token_id is None:
            warnings.warn(
                "pad_token_id is not set for the tokenizer. Using eos_token_id as pad_token_id."
            )
            tokenizer.pad_token = tokenizer.eos_token
        tokenizer.padding_side = "left"
        self.tokenizer = tokenizer

        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.model.eval()
        self.model.to(device=device, dtype=torch_dtype)

        self.generate_kwargs = {
            "temperature": 0.5,
            "top_p": 0.92,
            "top_k": 0,
            "max_new_tokens": 512,
            "use_cache": True,
            "do_sample": True,
            "eos_token_id": self.tokenizer.eos_token_id,
            "pad_token_id": self.tokenizer.pad_token_id,
            "repetition_penalty": 1.1,  # 1.0 means no penalty, > 1.0 means penalty, 1.2 from CTRL paper
        }

    def format_instruction(self, instruction):
        return PROMPT_FOR_GENERATION_FORMAT.format(instruction=instruction)

    def __call__(
        self, instruction: str, **generate_kwargs: Dict[str, Any]
    ) -> Tuple[str, str, float]:
        s = PROMPT_FOR_GENERATION_FORMAT.format(instruction=instruction)
        input_ids = self.tokenizer(s, return_tensors="pt").input_ids
        input_ids = input_ids.to(self.model.device)
        gkw = {**self.generate_kwargs, **generate_kwargs}
        with torch.no_grad():
            output_ids = self.model.generate(input_ids, **gkw)
        # Slice the output_ids tensor to get only new tokens
        new_tokens = output_ids[0, len(input_ids[0]) :]
        output_text = self.tokenizer.decode(new_tokens, skip_special_tokens=True)
        return output_text

class StopOnTokens(StoppingCriteria):
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
        for stop_id in stop_token_ids:
            if input_ids[0][-1] == stop_id:
                return True
        return False

model_path = '/home/work/data/models/generate_models/mosaicml_mpt-7b-instruct'
generate = InstructionTextGenerationPipeline(
    model_path,
    torch_dtype=torch.float16,
    trust_remote_code=True,
)
stop_token_ids = generate.tokenizer.convert_tokens_to_ids(["<|endoftext|>"])


def generate_text(instruction, temperature=0.1, top_p=1, top_k=0, max_new_tokens=512):
    if temperature < 0.1:
        temperature = 0.0
        do_sample = False
    else:
        do_sample = True
    
    stop = StopOnTokens()
    gkw = {
        **generate.generate_kwargs,
        **{
            "max_new_tokens": max_new_tokens,
            "temperature": temperature,
            "do_sample": do_sample,
            "top_p": top_p,
            "top_k": top_k,
            "stopping_criteria": StoppingCriteriaList([stop]),
        },
    }
    
    response = generate(instruction, **gkw)

    return response
```

**Comment by samhavens** [comment]
It is not the case that 1 number == 1 token. If you are putting that string into the template, that is 3,249 input tokens. ALiBi allows length extrapolation, but the quality does fall off as you go beyond the training length.

Another factor here is that instruction tuning examples in the wild are short; the ones we used were on average much shorter than the 2048 max, and we have found that performance starts to degrade around 1500 tokens.

We are working on curating longer instruction following data.

As a final note, putting the task at the beginning of the prompt makes it harder, I would try

```python
inp = PROMPT_FOR_GENERATION_FORMAT.format(instruction=' '.join([str(i) for i in range(2000)]) + "\nplease tell me what is the next number?")
```

Not certain, but the model has a better chance with that
